src/darwin/Framework/CHIP/MTRDevice_Concrete.mm

```diff
@@ -131,6 +131,67 @@ - (void)_deviceInternalStateChanged:(MTRDevice *)device;
 
 } // anonymous namespace
 
+#pragma mark - MTRDeviceMatterCPPObjectsHolder
+
+// Class to hold C++ objects that can only be manipulated on the Matter queue
+@interface MTRDeviceMatterCPPObjectsHolder : NSObject
+@property (nonatomic, readonly) ReadClient * readClient;
+@property (nonatomic, readonly) SubscriptionCallback * subscriptionCallback; // valid when and only when currentReadClient is valid
+- (void)setReadClient:(ReadClient * _Nullable)readClient subscriptionCallback:(SubscriptionCallback * _Nullable)subscriptionCallback;
+- (void)clearReadClientAndDeleteSubscriptionCallback;
+@end
+
+@implementation MTRDeviceMatterCPPObjectsHolder
+@synthesize readClient = _readClient;
+@synthesize subscriptionCallback = _subscriptionCallback;
+- (ReadClient *)readClient
+{
+    assertChipStackLockedByCurrentThread();
+    @synchronized(self) {
```

Comment:
I don't understand this `@synchronized` (or the others ones in this class).  We are already asserting we are running on a specific queue, no?  What is there to synchronize against?

Reply:
See https://github.com/project-chip/connectedhomeip/issues/37420